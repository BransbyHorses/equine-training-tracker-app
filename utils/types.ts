export interface Disruption {
	id: number;
	reason: { string: DisruptionCode; id: number };
	startDate: string;
	endDate?: string;
}

export interface DisruptionSimplified {
	id: number;
	name: string;
}

export interface Status {
	id: number;
	name: string;
	categorisedAsTraining: boolean;
}

export interface Equine {
	id: number;
	name: string;
	yard: Yard;
	equineStatus: EquineStatus;
	trainingProgrammes: TrainingProgramme[];
	learnerType: LearnerType;
	healthAndSafetyFlags: HealthAndSafetyFlag[];
	disruptions: Disruption[];
}

export interface EquineStatus {
	string:
		| "Awaiting Training"
		| "In Training"
		| "Returned To Owner"
		| "Rehomed"
		| "Euthanised"
		| "Other";
	id: 1 | 2 | 3 | 4 | 5 | 6;
	categorisedAsTraining: boolean;
}

export interface LearnerType {
	id: number;
	name: string;
}

export interface Skill {
	id: number;
	name: string;
}

export interface SkillProgressRecord {
	id: number;
	skill: Skill;
	trainingProgramme: TrainingProgramme;
	progressCode: { string: ProgressCode };
	startDate: string;
	endDate: string;
	time: number;
}

export interface SkillTrainingSession {
	id: number;
	date: string;
	trainingProgramme: TrainingProgramme;
	skill: Skill;
	environment: TrainingEnvironment;
	progressCode: { string: string };
	trainingMethod: TrainingMethod;
	trainingTime: number;
	notes: string;
}

export interface TrainingCategory {
	id: number;
	name: string;
}

export interface TrainingEnvironment {
	id: number;
	name: string;
}

export interface TrainingMethod {
	id: number;
	name: string;
}

export interface TrainingProgramme {
	id: number;
	trainingCategory: TrainingCategory;
	equine: Equine;
	skillProgressRecords: SkillProgressRecord[];
	skillTrainingSessions: SkillTrainingSession[];
	startDate: string;
	endDate: string;
}

export interface Yard {
	id: number;
	name: string;
}

export interface HealthAndSafetyFlag {
	id: number;
	content: string;
	dateCreated: string;
}

export enum ProgressCode {
	"Not able",
	"Limited",
	"Ok",
	"Confident",
}

export enum DisruptionCode {
	"Vetinary Review",
	"Team Low",
	"Weather",
	"Yard Busy",
	"Equine Wellbeing",
	"Other",
}
