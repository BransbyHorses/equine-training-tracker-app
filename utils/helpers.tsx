import { Equine, TrainingProgramme, SkillTrainingSession } from "./types";



export const findCurrentTrainingProgramme = (
	trainingProgrammes?: TrainingProgramme[]
) => {
	if (!trainingProgrammes) {
		return null;
	}

	const filteredTrainingProgrammes = trainingProgrammes.filter(
		(t) => !t.endDate
	);
	return filteredTrainingProgrammes.length === 0
		? null
		: filteredTrainingProgrammes[0];
};

export const convertDateToString = (dateString?: string) => {
	if (!dateString) return "";

	const date = new Date(dateString);
	return date.toLocaleString("en-GB", {
		year: "numeric",
		month: "long",
		day: "numeric",
	});
};

export const findLastTrainingSession = (
	skillTrainingSessions?: SkillTrainingSession[]
) => {
	if (!skillTrainingSessions || skillTrainingSessions.length === 0) return null;

	const sortedRecordsByDate = skillTrainingSessions.sort((a, b) => {
		return new Date(b.date).getTime() - new Date(a.date).getTime();
	});

	return sortedRecordsByDate[0];
};

export const generateTodaysDate = () => {
	let date = new Date();
	let today = date.getFullYear() + "-" + padZero((date.getMonth() + 1), 2) + "-" + padZero(date.getDate(), 2);
	return today;
}

const padZero = (num: number, pad: number) => num.toString().padStart(pad, '0');

export const saveData = async (data:any, path:string, method:string) => {
	console.log(data);
	await fetch(`${process.env.NEXT_PUBLIC_URL}data/${path}`, {
		method: method,
		headers: {
			"Content-Type": "application/json",
		},
		body: JSON.stringify(data),
	})
		.then((response) => response.json())
		.then((data) => {
			console.log("success");
		})
		.catch((rejected: any) => {
			console.log(rejected);
		});
}
