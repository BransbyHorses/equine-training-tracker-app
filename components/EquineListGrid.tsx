import Link from "next/link";
import { Equine } from "../utils/types";
import Box from "@mui/material/Box";
import Grid from "@mui/material/Grid";
import Typography from "@mui/material/Typography";
import IconButton from "@mui/material/IconButton";
import KeyboardArrowRightIcon from "@mui/icons-material/KeyboardArrowRight";
import { useEffect, useState } from "react";
import { useRouter } from "next/router";
import PaginationContainer from "./PaginationContainer";

const EquineListGrid = ({ equines }: { equines: Equine[] }) => {
	const router = useRouter();
	const [isAdmin, setIsAdmin] = useState<boolean>();

	useEffect(() => {
		router.pathname.includes("admin") ? setIsAdmin(true) : setIsAdmin(false);
	}, []);

	return (
		<>
			<PaginationContainer count={equines.length}>
				{equines
					?.sort((a, b) => {
						if (a.name < b.name) return -1;
						if (a.name > b.name) return 1;
						return 0;
					})
					.map((equine, i, arr) => {
						return (
							<Link
								href={
									isAdmin
										? `/admin/equines/${equine.id}`
										: `/equines/${equine.id}`
								}
								key={equine.id}
							>
								<Box
									p={2}
									sx={{
										display: "flex",
										justifyContent: "space-between",
										alignItems: "center",
										borderBottom:
											i === arr.length - 1 ? "0" : "0.5px solid lightGray",
										cursor: "pointer",
										backgroundColor: "white",
									}}
								>
									<Grid container>
										<Grid item xs={6} md={4} lg={3}>
											<Typography
												variant="h6"
												fontWeight={500}
												color="primary.light"
											>
												{equine.name}
											</Typography>
										</Grid>
										{equine.yard ? (
											<Grid
												item
												xs={5}
												md={4}
												lg={3}
												sx={{ display: "flex", alignItems: "center" }}
											>
												<Typography color="gray">{equine.yard.name}</Typography>
											</Grid>
										) : (
											<></>
										)}
									</Grid>
									<IconButton>
										<KeyboardArrowRightIcon fontSize="large" />
									</IconButton>
								</Box>
							</Link>
						);
					})}
			</PaginationContainer>
		</>
	);
};
export default EquineListGrid;
