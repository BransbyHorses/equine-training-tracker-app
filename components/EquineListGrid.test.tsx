import renderer from "react-test-renderer";
import "@testing-library/jest-dom";
import EquineListGrid from "./EquineListGrid";
import { Equine } from "../utils/types";
import { render, screen, fireEvent } from "@testing-library/react";
import { createMockRouter } from "../utils/test-utils/mockNextRouter";
import { RouterContext } from "next/dist/shared/lib/router-context";

const equines: Equine[] = [
	{
		id: 1,
		name: "Test Equine",
		yard: { id: 1, name: "Test Yard" },
		equineStatus: {
			string: "In Training",
			id: 1,
			categorisedAsTraining: true,
		},
		trainingProgrammes: [],
		learnerType: { id: 1, name: "Test Learner Type" },
		healthAndSafetyFlags: [],
		disruptions: [],
	},
];

describe("EquineListGrid", () => {
	it("renders with all props", () => {
		const tree = renderer.create(<EquineListGrid equines={equines} />);
		expect(tree).toMatchSnapshot();
	});

	it("displays non admin address when on index page", () => {
		const router = createMockRouter({});
		const expectedParams = [
			"/equines/1",
			"/equines/1",
			{ scroll: undefined, shallow: undefined, locale: undefined },
		];
		render(
			<RouterContext.Provider value={router}>
				<EquineListGrid equines={equines} />
			</RouterContext.Provider>
		);
		fireEvent.click(screen.getByTestId("KeyboardArrowRightIcon"));
		expect(router.push).toHaveBeenCalledWith(...expectedParams);
	});

	it("displays links to admin pages when at admin address", () => {
		const router = createMockRouter({ pathname: "/admin/equines" });
		const expectedParams = [
			"/admin/equines/1",
			"/admin/equines/1",
			{ scroll: undefined, shallow: undefined, locale: undefined },
		];
		render(
			<RouterContext.Provider value={router}>
				<EquineListGrid equines={equines} />
			</RouterContext.Provider>
		);
		fireEvent.click(screen.getByTestId("KeyboardArrowRightIcon"));
		expect(router.push).toHaveBeenCalledWith(...expectedParams);
	});
});
