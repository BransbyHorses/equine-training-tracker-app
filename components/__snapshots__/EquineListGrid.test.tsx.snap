// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EquineListGrid renders with all props 1`] = `
[
  <div
    className="MuiBox-root css-sp4511"
    onClick={[Function]}
    onMouseEnter={[Function]}
  >
    <div
      className="MuiGrid-root MuiGrid-container css-11lq3yg-MuiGrid-root"
    >
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-1qkef4z-MuiGrid-root"
      >
        <h6
          className="MuiTypography-root MuiTypography-h6 css-pl9v5i-MuiTypography-root"
        >
          Test Equine
        </h6>
      </div>
      <div
        className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-5 MuiGrid-grid-md-4 MuiGrid-grid-lg-3 css-k8h3ww-MuiGrid-root"
      >
        <p
          className="MuiTypography-root MuiTypography-body1 css-1rfn1w6-MuiTypography-root"
        >
          Test Yard
        </p>
      </div>
    </div>
    <button
      className="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeMedium css-78trlr-MuiButtonBase-root-MuiIconButton-root"
      disabled={false}
      onBlur={[Function]}
      onContextMenu={[Function]}
      onDragLeave={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      onMouseDown={[Function]}
      onMouseLeave={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      onTouchMove={[Function]}
      onTouchStart={[Function]}
      tabIndex={0}
      type="button"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeLarge css-tzssek-MuiSvgIcon-root"
        data-testid="KeyboardArrowRightIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M8.59 16.59 13.17 12 8.59 7.41 10 6l6 6-6 6-1.41-1.41z"
        />
      </svg>
    </button>
  </div>,
  <div
    className="MuiBox-root css-1bvc4cc"
  >
    <nav
      aria-label="pagination navigation"
      className="MuiPagination-root MuiPagination-text css-pg6uc-MuiPagination-root"
    >
      <ul
        className="MuiPagination-ul css-wjh20t-MuiPagination-ul"
      >
        <li>
          <button
            aria-label="Go to previous page"
            className="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
              data-testid="NavigateBeforeIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M15.41 7.41L14 6l-6 6 6 6 1.41-1.41L10.83 12z"
              />
            </svg>
          </button>
        </li>
        <li>
          <button
            aria-current="true"
            aria-label="page 1"
            className="MuiButtonBase-root MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-selected MuiPaginationItem-page css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            1
          </button>
        </li>
        <li>
          <button
            aria-label="Go to next page"
            className="MuiButtonBase-root Mui-disabled MuiPaginationItem-root MuiPaginationItem-sizeMedium MuiPaginationItem-text MuiPaginationItem-circular Mui-disabled MuiPaginationItem-previousNext css-yuzg60-MuiButtonBase-root-MuiPaginationItem-root"
            disabled={true}
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={-1}
            type="button"
          >
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiPaginationItem-icon css-g2z002-MuiSvgIcon-root-MuiPaginationItem-icon"
              data-testid="NavigateNextIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 6L8.59 7.41 13.17 12l-4.58 4.59L10 18l6-6z"
              />
            </svg>
          </button>
        </li>
      </ul>
    </nav>
  </div>,
]
`;
